### PR TITLE
Project setup: eliminate the need to set the user.timezone VM flag #7520

### LINKED ISSUE #7520: Project setup: eliminate the need to set the user.timezone VM flag
The current way to ensure UTC time in the dev server is to set the VM flag `-Duser.timezone=UTC` in the dev server run configurations. This instruction is easy to miss and user-dependent. Instead, the time zone can be set in the application code itself run during server startup. This will enforce consistency and ease the setup process.

There is also one test, `StudentProfileAttributesTest`, that depends on the test environment being on UTC for it to pass. With this test patched, the entire test suite can be run in the user's local time zone without breaking, and it will no longer be necessary to have the VM flag in the test environment run configurations.

### PR DESCRIPTION
**Pull request by whipermr5**
Fixes #7520

**Outline of Solution**
- Set time zone to UTC in `ControllerServlet` init
  - Add `load-on-startup` flag to `ControllerServlet` in `web.xml` to ensure code runs at server startup
- Fix `StudentProfileAttributesTest` to be independent of system time zone
  - Fix `JsonUtils` to always read and write JSON dates in UTC, independent of system time zone (lightens reviewer's burden of checking time zone of date fields in JSON files in PRs that touch them)

### CODE REVIEW

File: src/main/java/teammates/ui/controller/ControllerServlet.java [hunk 1]
```diff
@@ -39,6 +41,11 @@
     private static final Logger log = Logger.getLogger();
 
     @Override
+    public void init() throws ServletException {
+        TimeHelper.setSystemTimeZoneIfRequired();
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/ControllerServlet.java]: Will other Servlets be fine without this?

File: src/main/java/teammates/common/util/JsonUtils.java [hunk 2]
```diff
@@ -23,8 +32,7 @@ private JsonUtils() {
      * Json file and also reformat the Json string in pretty-print format.
      */
     private static Gson getTeammatesGson() {
-        return new GsonBuilder().setDateFormat(DateFormat.FULL)
-                                .setDateFormat(Const.SystemParams.DEFAULT_DATE_TIME_FORMAT)
+        return new GsonBuilder().registerTypeAdapter(Date.class, new StandardDateAdapter())
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/JsonUtils.java]: Will this have any effect to legacy data (i.e. those existing before this PR)?

File: src/main/java/teammates/common/util/JsonUtils.java [hunk 3]
```diff
@@ -72,4 +80,31 @@ public static JsonElement parse(String json) {
...
+        StandardDateAdapter() {
+            dateFormat = new SimpleDateFormat(Const.SystemParams.DEFAULT_DATE_TIME_FORMAT);
+            dateFormat.setTimeZone(Const.SystemParams.TIME_ZONE);
+        }
+
+        @Override public synchronized JsonElement serialize(Date date, Type type, JsonSerializationContext context) {
+            return new JsonPrimitive(dateFormat.format(date));
+        }
+
+        @Override public synchronized Date deserialize(JsonElement element, Type type, JsonDeserializationContext context) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/JsonUtils.java]: Nit:
```java
@Override
public ... {
    ...
}
```
Similarly for `serialize`.

File: src/main/java/teammates/ui/controller/ControllerServlet.java (hunk 1 above)
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/ControllerServlet.java]: This servlet will be loaded before all the other servlets. The call affects the entire application (even the timestamp of the console log messages changes).

Comment (whipermr5) [inline src/main/java/teammates/ui/controller/ControllerServlet.java]: Example:
screen shot 2017-06-19 at 8 26 41 am

File: src/main/java/teammates/common/util/JsonUtils.java (hunk 2 above)
Comment (whipermr5) [inline src/main/java/teammates/common/util/JsonUtils.java]: There's no actual change to the expected format - `Const.SystemParams.DEFAULT_DATE_TIME_FORMAT` is still used. `setDateFormat(DateFormat.FULL)` previously didn't do anything as it was overwritten by `.setDateFormat(Const.SystemParams.DEFAULT_DATE_TIME_FORMAT)`.

Comment (whipermr5) [inline src/main/java/teammates/common/util/JsonUtils.java]: From their [docs](https://github.com/google/gson/blob/gson-parent-2.8.0/gson/src/main/java/com/google/gson/GsonBuilder.java#L391-L393):
> `Configures Gson to serialize {@code Date} objects according to the pattern provided. You can call this method or {@link #setDateFormat(int)} multiple times, but only the last invocation will be used to decide the serialization format.`

Comment (whipermr5) [inline src/main/java/teammates/common/util/JsonUtils.java]: Their default adapter code can be found here: https://github.com/google/gson/blob/gson-parent-2.8.0/gson/src/main/java/com/google/gson/DefaultDateTypeAdapter.java

Comment (whipermr5) [inline src/main/java/teammates/common/util/JsonUtils.java]: Renaming to `TeammatesDateAdapter` upon suggestion by @LiHaoTan.

File: docs/development.md [hunk 4]
```diff
@@ -150,17 +150,7 @@ You need a student account which can be created by instructors.
...
-* Specify timezone as a VM argument:
-  * Eclipse
-    * Go to the run configuration Eclipse created when you started the dev server (`Run → Run configurations ...` and select the appropriate one).
-    * Click on the `Arguments` tab and add `-Duser.timezone=UTC` to the `VM arguments` text box.
-    * Save the configuration for future use: Go to the `Common` tab (the last one) and make sure you have selected `Save as → Local file` and `Display in favorites menu → Run, Debug`.
-  * IntelliJ
-    * Go to `Run → Edit Configurations...` and select `Dev Server`.
-    * Add `-Duser.timezone=UTC` to the `VM options` text box. Click `OK`.
-* Start the server again using the run configuration you created in the previous step.
+Note that the dev server sets its time zone to UTC at startup.
```
Comment (LiHaoTan) [inline docs/development.md]: Should we make it more prominent like this?

**NOTE**
> The dev server sets its time zone to UTC at startup.

**Comment by whipermr5** [comment]
> Will there be gotchas like time zone being unexpectedly set to something else, or a sudden need for time zone to be set to something else?

The code only affects dev servers on non-UTC systems; there will be no effect on the live or staging servers as they all run on UTC. If we ever need to set to something else other than UTC, we can edit `Const.SystemParams.TIME_ZONE`.

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 5]
```diff
@@ -187,28 +202,33 @@ public static String formatTime12H(Date date) {
         if (date == null) {
             return "";
         }
-        Calendar c = Calendar.getInstance();
+        SimpleDateFormat sdf = null;
```
Comment (LiHaoTan) [inline src/main/java/teammates/common/util/TimeHelper.java]: I don't think we should reuse this variable.

Comment (LiHaoTan) [inline src/main/java/teammates/common/util/TimeHelper.java]: LOL ok

**Pull request merged**